include/pybind11/pytypes.h

```diff
@@ -1059,6 +1059,34 @@ inline str::str(const bytes& b) {
     m_ptr = obj.release().ptr();
 }
 
+/// \addtogroup pytypes
+/// @{
+class bytearray : public object {
+public:
+    PYBIND11_OBJECT(bytearray, object, PyByteArray_Check)
+
+    bytearray()
+        : object(PyByteArray_FromStringAndSize("", (ssize_t) 0), stolen_t{}) {
+        if (!m_ptr) pybind11_fail("Could not allocate bytearray object!");
+    }
+
+    bytearray(const char *c, size_t n)
+        : object(PyByteArray_FromStringAndSize(c, (ssize_t) n), stolen_t{}) {
+        if (!m_ptr) pybind11_fail("Could not allocate bytearray object!");
+    }
+
+    // Allow implicit conversion:
+    bytearray(const std::string &s) : bytearray(s.data(), s.size()) { }
+
+    operator std::string() const {
+        char *buffer = PyByteArray_AsString(m_ptr);
+        return std::string(buffer);
```

Comment:
1. Get the size as well, with `PyByteArray_GET_SIZE`, because `std::string{buffer, size}` is more efficient.
2. Do we need to `free(buffer)` after passing it off to the string?
3. Can `PyByteArray_AsString()` ever return `NULL`? Perhaps we need to guard against it.

Reply:
1. Done
2. I don't think we need to free the buffer since PyByteArray_AsString is not allocating anything it is just returning a pointer to already allocated bytearray as far as I understand.
3. PyByteArray_AsString() will return NULL only if m_ptr which it will never be as it is checked in the constructor. So in fact I am now using macro versions of the methods as documented here: https://docs.python.org/3/c-api/bytearray.html which do not check if m_ptr is NULL.